### DISCUSSION TITLE
Elsa 2.0 Why is that the workflow definition table is not populated while the instance table is populated

### DISCUSSION THREAD

**Discussion by AvtsVivek** [elsa-workflows/elsa-core#1058] [Q&A]
When [this simple workflow ](https://github.com/AvtsVivek/ElsaBugReports/tree/main/src/P20240PersistenceEfMsSql)is persisted, why is that the workflow definition is not persisted? We can see the instance is persisted.

The WorkflowDefinition table is not populated while the instance table is populated with Definition Id.

![image](https://user-images.githubusercontent.com/41255281/119655434-35d46700-be47-11eb-924c-ba1a7f2f89d4.png)

What should I do to make an entry to this table?

Below is the code.

    public class Program
    {
        private static async Task Main()
        {
            var services = new ServiceCollection()
                .AddElsa(options => options
                    .UseEntityFrameworkPersistence(ef =>
                    {
                        ef.UseSqlServer(@"Data Source=(localdb)\MSSQLLocalDB;Initial Catalog=Elsa20;Integrated Security=True;Connect Timeout=30;Encrypt=False;TrustServerCertificate=False;ApplicationIntent=ReadWrite;MultiSubnetFailover=False");
                    })
                    .AddConsoleActivities())
                .AddAutoMapperProfiles<Program>()
                .BuildServiceProvider();

            var startupRunner = services.GetRequiredService<IStartupRunner>();
            await startupRunner.StartupAsync();

            var workflowRunner = services.GetRequiredService<IBuildsAndStartsWorkflow>();

            var runWorkflowResult = await workflowRunner.BuildAndStartWorkflowAsync<HelloWorldPersistanceWorkflow>();

            var store = services.GetRequiredService<IWorkflowInstanceStore>();

            var count = await store.CountAsync(new WorkflowDefinitionIdSpecification(nameof(HelloWorldPersistanceWorkflow)));

            Console.WriteLine(count);

            var loadedWorkflowInstance = await store.FindByIdAsync(runWorkflowResult.WorkflowInstance.Id);
            Console.WriteLine(loadedWorkflowInstance);

And here is the workflow. The code is available [here](https://github.com/AvtsVivek/ElsaBugReports/tree/main/src/P20240PersistenceEfMsSql).

    public class HelloWorldPersistanceWorkflow : IWorkflow
    {
        public void Build(IWorkflowBuilder builder)
        {
            builder // You need to set this option for your workflow to persist
                .WithPersistenceBehavior(WorkflowPersistenceBehavior.WorkflowBurst)
                .WriteLine("Hello World!");
        }
    }

**Comment by sfmskywalker** [comment]
This is by design.

A workflow instance points to a **Workflow Blueprint** which themselves are not stored in the database.
Workflow Blueprints are stored in an in-memory service called the **Workflow Registry**.

The Workflow Registry is populated from various sources, such as:

* Workflow Definitions stored in the database.
* Workflow classes built with code (as in your example).

Workflow Instances are always stored in the database since they represent runtime information about the workflow. Workflow blueprints do not need to be stored, since they are "static" such as your workflow class.

That being said, I do have plans to unify the concept of "workflow definition" such that it does not matter whether the source is JSON-based or code-based as outlined in #569

When that is implemented, workflow classes will be compiled into an assembly, which is then stored in the database as well in the form of a Workflow Definition. The Data column will then simply store an array of bytes instead of JSON in the case of workflows created by the designer. There will then be an additional column, probably called something like "Content Type".

It also solves the issue of versioning workflow classes.
